### PR TITLE
[ERE-2181] Default consent video language to user's language

### PR DESCRIPTION
**Pull request by Lexiel46**

### CODE REVIEW

**Comment by Lexiel46** [comment]
To test the function, below codes should be added to the "Information text" box in a consent section. In addition, currently we have only one text track file so the src link is not changed based on language.
```django
{% for lang in all_available_langs %}
{% if current_language.code == lang.0 or current_language.name_local == lang.1 %}
<track
      label="{{ lang.1 }}"
      kind="subtitles"
      srclang="{{ lang.0 }}"
src="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/qut_subs.vtt" default />
{% else %}
<track
      label="{{ lang.1 }}"
      kind="subtitles"
      srclang="{{ lang.0 }}"
src="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/qut_subs.vtt" />
{% endif %}
{% endfor %}
```

**Comment by ppettitau** [comment]
What happens if the user selects a caption language that exists in the registry, but has no associated subtitles?
For example, I configured my local registry to support French `fr`, it generated a track for "Francais", but then when I selected it the Caption Language option disappeared. Is this expected?

_Disclaimer_ I tested the raw HTML output in the Consent Section in jsfiddle because my CSP configuration doesn't allow the Angelman video to be rendered.

1) Before selecting Francais:
image

image

2) After selecting Francais:
image

File: rdrf/rdrf/templatetags/get_info_text_expression.py [hunk 1]
```diff
@@ -21,7 +23,13 @@ def get_info_text_expression(fields):
         try:
             consent_section_model = ConsentSection.objects.get(pk=consent_section_model_pk)
             if consent_section_model.information_text:
-                return _(consent_section_model.information_text)
+                information_text = Template(consent_section_model.information_text)
+                context = Context({
+                    'all_available_langs': settings.LANGUAGES,
+                    'current_language': current_lang,
```
Comment (ppettitau) [inline rdrf/rdrf/templatetags/get_info_text_expression.py]: Consider using consistent naming for variables. Where possible, use the same naming to represent "language" e.g. always either use "language" or "lang", but not a mix of both.
* all_available_langs
* current_language
* current_lang

Personally, I prefer the non-abbreviated form "language".

**Comment by ppettitau** [comment]
> What happens if the user selects a caption language that exists in the registry, but has no associated subtitles? For example, I configured my local registry to support French `fr`, it generated a track for "Francais", but then when I selected it the Caption Language option disappeared. Is this expected?
> 
> _Disclaimer_ I tested the raw HTML output in the Consent Section in jsfiddle because my CSP configuration doesn't allow the Angelman video to be rendered.
> 
> 1. Before selecting Francais:
>    image
> 
> image
> 
> 2. After selecting Francais:
>    image

I retried using proper configuration for enabling the video, and enabling the languages. It renders fine in TRRF. Sorry for the added confusion! Not sure exactly what happened the first time, but possibly didn't include the `track` rendering code as expected in the Consent Section info text.

**Comment by Lexiel46** [comment]
Changing the variable name would lead to the change of code entering into the information text box.
Please use `{% for lang in all_available_languages %}` to replace `{% for lang in all_available_langs %}`

**Review by ppettitau — APPROVED** [review]
Great work, Lexie :+1:

**Comment by ScriptSmith** [comment]
If you use the following env variables, you can test it locally:

```
LANGUAGES="en pseudo af ar-AR zh-CN zh-TW cs fr de he hi hu id it ms pl pt ro ru es sv tr ur vi"
ENABLE_VIDEO_HOSTING=1
VIDEO_HOSTING_BASE_URL="https://d4d7dpr65gp9x.cloudfront.net/"
```

**Review by ScriptSmith — CHANGES_REQUESTED** [review]
Python part is great, but we need to test using a different file for different languages. I would suggest adding a file to the bucket with the same name as the language code, and using that in the template.

And as @ppettitau alluded to, missing a subtitle file with the above template would cause a 403 and leave a non-working button in the video. So we need to ensure it's available for all supported languages in the registry, or perhaps use JS to check the file exists first.

When there are a lot of languages enabled, the scrollbar in the list of subtitles is quite subtle which is annoying. Might be worth tweaking the style to make it more visible?

Also noticed that there was a 'Not same-origin' error for the default track when loading the page. Adding the `crossorigin="anonymous"` attribute to the video seemed to fix this.

**Comment by Lexiel46** [comment]
Different from previous method, there's no code to be added to template now. Instead, please at below js code to the information text box and test.
```
<script>
       var available_text_tracks = ['en', 'fr']; // Manage this array manually by checking the available text tracks in aws
      $(document).ready(function() {
          var player = videojs('TestVideo');
          var all_available_languages = {{ all_available_languages | safe }};
          var current_language = {{ current_language | safe }};
          player.ready(() => {
              for (let i = 0; i < all_available_languages.length; i++) {
                  let language = all_available_languages[i];
                  let src = `https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/${language[0]}.vtt`;
                  if (available_text_tracks.includes(language[0])) {
                       player.addRemoteTextTrack({src: src, kind:'subtitles', srclang: language[0], label: language[1], default: current_language[0] == language[0] || current_language[1] == language[1]});
                  }
              }
          });
      });
</script>
```

**Comment by ppettitau** [comment]
> estVideo');
>           var all_available_lang

This is very nitpicky, particularly as this JS isn't repository code, but a couple of things:
* Prefer using `const` and `let` over `var`
* Use `const` unless you are reassigning the value of the variable. 
* `src`, `current_language` is defined outside of the if condition, but it should only be assigned if we are going to use it, so move it inside the `if`.
* I don't think we gain anything by declaring `player`, this could be a one-liner

**Comment by Lexiel46** [comment]
Updated the js code as below. I think we should keep the declaration of `player` because both `player.ready` and `player.addRemoteTextTrack` require that. What do you think? @ppettitau 
```
<script>
       const available_text_tracks = ['en', 'fr']; // Manage this array manually by checking the available text tracks in aws
      $(document).ready(function() {
          const player = videojs('TestVideo');
          const all_available_languages = {{ all_available_languages | safe }};
          const current_language = {{ current_language | safe }};
          player.ready(() => {
              for (let i = 0; i < all_available_languages.length; i++) {
                  const language = all_available_languages[i];
                  if (available_text_tracks.includes(language[0])) {
                       const src = `https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/${language[0]}.vtt`;
                       player.addRemoteTextTrack({src: src, kind:'subtitles', srclang: language[0], label: language[1], default: current_language[0] == language[0] || current_language[1] == language[1]});
                  }
              }
          });
      });
</script>
```

**Comment by ppettitau** [comment]
> Updated the js code as below. I think we should keep the declaration of `player` because both `player.ready` and `player.addRemoteTextTrack` require that. What do you think? @ppettitau
> 
> ```
> <script>
>        const available_text_tracks = ['en', 'fr']; // Manage this array manually by checking the available text tracks in aws
>       $(document).ready(function() {
>           const player = videojs('TestVideo');
>           const all_available_languages = {{ all_available_languages | safe }};
>           const current_language = {{ current_language | safe }};
>           player.ready(() => {
>               for (let i = 0; i < all_available_languages.length; i++) {
>                   const language = all_available_languages[i];
>                   if (available_text_tracks.includes(language[0])) {
>                        const src = `https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/${language[0]}.vtt`;
>                        player.addRemoteTextTrack({src: src, kind:'subtitles', srclang: language[0], label: language[1], default: current_language[0] == language[0] || current_language[1] == language[1]});
>                   }
>               }
>           });
>       });
> </script>
> ```

Looks good :+1: And yes, that makes sense - I missed that reference to the player.

**Comment by ScriptSmith** [comment]
More nitpicky things would be to use for...of ,destructuring assignment, and `Set`s like so:

```diff
 <script>
-    const available_text_tracks = ['en', 'fr']; // Manage this array manually by checking the available text tracks in aws
+    const available_text_tracks = new Set(['en', 'fr']); // Manage this array manually by checking the available text tracks in aws
     $(document).ready(function () {
         const player = videojs('ConsentVideo');
         const all_available_languages = {{ all_available_languages | safe }};
-        const current_language = {{ current_language | safe }};
+        const [current_language_code, current_language_label] = {{ current_language | safe }};
         player.ready(() => {
-            for (let i = 0; i < all_available_languages.length; i++) {
-                const language = all_available_languages[i];
-                if (available_text_tracks.includes(language[0])) {
-                    const src = `https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/${language[0]}.vtt`;
+            for (const [language_code, language_label] of all_available_languages) {
+                if (available_text_tracks.has(language_code)) {
+                    const src = `https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/${language_code}.vtt`;
                     player.addRemoteTextTrack({
                         src: src, 
                         kind: 'subtitles',
-                        srclang: language[0],
-                        label: language[1],
-                        default: current_language[0] == language[0] || current_language[1] == language[1]
+                        srclang: language_code,
+                        label: language_label,
+                        default: current_language_code == language_code || current_language_label == language_label
                     });
                 }
             }
```

I think the `||` in the `default` property is redundant too, as code should be enough for the condition.

Here's my version of the video with all the updates:
```django
<video id="ConsentVideo" class="video-js" controls preload="auto"
    poster="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/poster/consent_20220629.0000001.jpg"
    data-setup='{"fluid": true}'>
    <source src="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/cmaf/consent_20220629.m3u8"
        type="application/x-mpegURL" />
    <source src="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/hls/consent_20220629.m3u8"
        type="application/x-mpegURL" />
    <source src="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/hls/consent_20220629.m3u8"
        type="application/x-mpegURL" />
    <source
        src="https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/mp4/consent_20220629_avc_720p.mp4"
        type="video/mp4" />
    <source src="angelman/01G6PFFT70CMQR8VS9G6W8N94C/mp4/consent_20220629_avc_480p.mp4" type="video/mp4" />
    <p class="vjs-no-js">
        To view this video please enable JavaScript, and consider upgrading to a
        web browser that
        <a href="https://videojs.com/html5-video-support/" target="_blank">supports HTML5 video</a>
    </p>
</video>

<script>
    const available_text_tracks = new Set(['en', 'fr']); // Manage this array manually by checking the available text tracks in aws
    $(document).ready(function () {
        const player = videojs('ConsentVideo');
        const all_available_languages = {{ all_available_languages | safe }};
        const [current_language_code, current_language_label] = {{ current_language | safe }};
        player.ready(() => {
            for (const [language_code, language_label] of all_available_languages) {
                if (available_text_tracks.has(language_code)) {
                    const src = `https://d4d7dpr65gp9x.cloudfront.net/angelman/01G6PFFT70CMQR8VS9G6W8N94C/subtitle/${language_code}.vtt`;
                    player.addRemoteTextTrack({
                        src: src, 
                        kind: 'subtitles',
                        srclang: language_code,
                        label: language_label,
                        default: current_language_code == language_code || current_language_label == language_label
                    });
                }
            }
        });
    });
</script>
```

**Comment by Lexiel46** [comment]
The `||` in the `default` property is for checking Chinese characters as our project is using `zh-CN: 简体中文` and `zh-TW: 繁體中文` for simplified Chinese and traditional Chinese while those code in Django are `zh-hans: 简体中文` and `zh-hant: 繁體中文` respectively. Hence, checking language code only is not enough in this case.

**Review by ScriptSmith — APPROVED** [review]

**Pull request merged**